### PR TITLE
Shwap

### PR DESCRIPTION
**Pull request by Wondertan**
Prototype of the new storage for edses - edsfile and optimized roundtrip less ipldv2 protocol

TODO:
- edsfile
	- [x] Blockstore over EDSFile
	- [x] Support for Col sample reads
	- [x] ODS only mode
		- [ ] ~Support for Axis, Share and ShareWithProof~
	- [x] Measure size diff
	- [x] Delete EDS mode
	- [x] GetSharesByNamespace (assigned to @walldiss) 
	- [ ] File streaming (Write / Read)
	- [ ] Concurrency safety for file  
- edsStore	
	- [ ] Base implementation
	- [ ] Caching
	- [ ] Migrate shrex server to new storage
	- [ ] Store getter 
- ipldv2
	- [x] Define Sample type and refactor the protocol and the Hasher
	- [x] Migrate to proto
	- [x] Row Sampling
	- [ ] ~Versioning and backwards compatibility with old protocol~
	- [x] Use heights instead of DataHash to identify block 
	- [x] Hash requested Axis roots to shorten requests
	- [ ] Retriever over v2 (IPLDv2 reconstruction getter) (assigned to @walldiss)
	- [ ] Metrics 
	- [x] IPLDv2 row based getter
		- [x] Getter should support requesting batch of shares by idx at once 
		- [x] Store only samples by default
		- [x] Bitswap Sessions 
	- [ ] ~IPLDv2 Server over LN storage (blockservice)~ Will work as is 
	- [ ] Migrate actual sampling logic over GetShares #2984 
	
Stretch Goals:
* Range of shares Multihash(by extending SampleID with additional uint16 that defines the end idx of range)
* Blob requests. BlobID{DataID, Comm} -> Blob{BlobShares, ProofsToSubroot}
* Requesting proofs only for every request
* Reconsider DHT discovery.

### CODE REVIEW

**Review by derrandz — COMMENTED** [review]
I love the direction this is taking, this is so good to look at.

I think this would better answer to the needs specified in https://github.com/celestiaorg/celestia-node/issues/2618 and would deprecate the gymnastics of reading from the car file done in https://github.com/celestiaorg/celestia-node/issues/2618

(_commenting for reference_)

**Comment by Wondertan** [comment]
@derrandz, very true

File: share/eds/file.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+	offset := int64(idx)*shrLn + HeaderSize
+	shr := make(share.Share, shrLn)
+	if _, err := f.fl.ReadAt(shr, offset); err != nil {
+		return nil, err
+	}
+	return shr, nil
+}
+
+func (f *File) ShareWithProof(idx int, axis rsmt2d.Axis) (share.Share, nmt.Proof, error) {
```
Comment (derrandz) [inline share/eds/file.go:120]: [feature request]: Having `AxisWithProof` will be super useful for search operations to retrieve all rows with their proofs and roots to be able to filter out for specific shares and their proofs.

**Review by derrandz — COMMENTED** [review]
a small feature request to make #2618 easier to implement

File: share/eds/file.go (hunk 1 above)
Comment (Wondertan) [reply share/eds/file.go:120]: Yep, wanted to make it

Comment (Wondertan) [reply share/eds/file.go:120]: When you get the whole axis, there is no proof.

File: share/shwap/pb/shwap_pb.proto [hunk 2]
```diff
@@ -0,0 +1,26 @@
+syntax = "proto3";
+
```
Comment (fl0rek) [inline share/shwap/pb/shwap_pb.proto:2]:
```suggestion

package share.p2p.shwap;
```
Or was package specifier ommited for some reason?

Comment (Wondertan) [reply share/shwap/pb/shwap_pb.proto:2]: It's not necessary from my understanding

**Comment by fl0rek** [comment]
I don't think it's necessary but  other protobuf files specify the
namespace,so we expected this one too

File: share/shwap/pb/shwap_pb.proto (hunk 2 above)
Comment (fl0rek) [reply share/shwap/pb/shwap_pb.proto:2]: I don't think it's necessary, but other proto files specify the namespace so I expected this one too

Comment (Wondertan) [reply share/shwap/pb/shwap_pb.proto:2]: Ok. Also, note that the name of the pkg is more like an implementation detail. The formal spec won't contain the pkg name in its proto file.

**Comment by Wondertan** [comment]
We merged #3675, so can close this one

**Pull request closed without merging**